docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md

```diff
@@ -614,6 +614,434 @@ share them with us at [#dev-support on our Community Slack channel][slack-devsup
+    let contracts = Solc::default().compile_source(&contract)?;
+    let c = contracts.get(&contract, &contract_name).unwrap();
+    let abi = c.abi.unwrap().to_owned();
+    let bytecode = c.bytecode().unwrap().to_owned();
+
+    // save the ABI to a file in order to load it in the call script
+    let abi_json = serde_json::to_string_pretty(&abi)?;
+    let abi_json_file_path = "/path/to/Greeter.abi";
+    let mut abi_json_file_handler = File::create(abi_json_file_path)?;
+    write!(abi_json_file_handler, "{}", abi_json)?;
+
+    // set up emerald-related variables (blockchain url, wallet, client)
+    let emerald_ip = std::env::var("EMERALD_IP").unwrap_or_else(|_| "127.0.0.1".into());
+    let emerald_uri = format!("http://{}:8545", emerald_ip);
+    let private_key = std::env::var("PRIVATE_KEY").expect("PRIVATE_KEY env var is mandatory");
+    let private_key = private_key.trim_start_matches("0x");
+    let provider = Provider::<Http>::try_from(emerald_uri)?.interval(Duration::from_secs(7));
+    let chain_id = provider.get_chainid().await?.as_u64();
+    let wallet = private_key.parse::<LocalWallet>()?.with_chain_id(chain_id);
+    println!("wallet public address is {:?}", wallet.address());
+    let client = Arc::new(SignerMiddleware::new(provider, wallet));
+
+    // deploy the contract
+    let factory = ContractFactory::new(abi, bytecode, client);
+    let contract = factory
+        .deploy("HELLO FROM RUST!".to_string())?
+        .legacy()
+        .confirmations(0usize)
+        .send()
+        .await?;
+    println!("Contract address: {:?}", contract.address());
+
+    Ok(())
+}
+```
+
+In case you don't want to trigger `solc` from rust, you can also compile the bytecode directly
+from bash by running `solc --abi --bin -o . Greeter.sol` and after that loading the ABI and the bytecode
+from the files:
+
+```ts
+    let abi: Abi =
+        serde_json::from_str(&fs::read_to_string("/path/to/Greeter.abi").unwrap())?;
+    let bytecode_string = fs::read_to_string("/path/to/Greeter.bin")?;
+    let bytecode = hex::decode(bytecode_string)?;
+    let bytecode = ethers_core::types::Bytes::from(bytecode);
+```
+
+After deploying the contract you should get a contract address that looks like this:
+`0xe0b863149d323464ecbf09048410d8c19994047d`
+
+### Call smart contracts from Rust
+
+The `Cargo.toml` required to call the smart contract is almost the same. You just need
+`serde` instead of `hex`:
+
+```
+[package]
+name = "sol-call"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+ethers = { git = "https://github.com/gakonst/ethers-rs/" }
+eyre = "0.6.8"
+serde_json = "1.0.81"
+serde = "1.0.137"
+tokio = { version = "1.17", features = [ "rt", "macros" ] }
+```
+
+In order to call the contract, you will need to load the ABI that you created earlier.
+You will still need the environment variable `PRIVATE_KEY` and in case your Emerald
+server is remote, you will also need to specify the `EMERALD_IP`. Apart from this, you 
```

Comment:
Why `EMERALD_IP`? You probably want to use host instead? Also, this is not really emerald-specific setting, so I would name this env var as `WEB3_HOST` instead.

Reply:
> Why `EMERALD_IP`? You probably want to use host instead?

This depends on the setup. On the specific setup I used to build the demo, I had an Emerald container running on a different IP. Host names were not added to that specific setup, though I agree host names are a requirement for prod setup.

> Also, this is not really emerald-specific setting, so I would name this env var as `WEB3_HOST` instead.

I suggest naming the env var `WEB3_HOST_URL`, so that the user can also specify `http` or `https` and the port. This makes a lot more sense than `EMERALD_IP`.